### PR TITLE
[Pulsar-Client] Add Producer Builder API Numeric Properties Validations

### PR DESCRIPTION
**Pull request by erenavsarogullari**
### Motivation
This PR proposes `Producer Builder API` numeric properties validation protecting `negative access` (due to Public API).

### Test
New UTs are added to `ProducerBuilderImplTest`

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java [hunk 1]
```diff
@@ -213,18 +214,21 @@ private ProducerBuilderImpl(PulsarClientImpl client, ProducerConfigurationData c
 
     @Override
     public ProducerBuilder<T> batchingMaxPublishDelay(long batchDelay, @NonNull TimeUnit timeUnit) {
+        checkArgument(batchDelay >= 0, "batchDelay needs to be >= 0");
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Why this can’t be negative?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java [hunk 2]
```diff
@@ -213,18 +214,21 @@ private ProducerBuilderImpl(PulsarClientImpl client, ProducerConfigurationData c
...
     @Override
     public ProducerBuilder<T> batchingMaxMessages(int batchMessagesMaxMessagesPerBatch) {
+        checkArgument(batchMessagesMaxMessagesPerBatch >= 0, "batchMessagesMaxMessagesPerBatch needs to be >= 0");
         conf.setBatchingMaxMessages(batchMessagesMaxMessagesPerBatch);
         return this;
     }
 
     @Override
     public ProducerBuilder<T> initialSequenceId(long initialSequenceId) {
+        checkArgument(initialSequenceId >= 0, "initialSequenceId needs to be >= 0");
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Why this can’t be negative?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java [hunk 3]
```diff
@@ -213,18 +214,21 @@ private ProducerBuilderImpl(PulsarClientImpl client, ProducerConfigurationData c
...
     @Override
     public ProducerBuilder<T> batchingMaxPublishDelay(long batchDelay, @NonNull TimeUnit timeUnit) {
+        checkArgument(batchDelay >= 0, "batchDelay needs to be >= 0");
         conf.setBatchingMaxPublishDelayMicros(timeUnit.toMicros(batchDelay));
         return this;
     }
 
     @Override
     public ProducerBuilder<T> batchingMaxMessages(int batchMessagesMaxMessagesPerBatch) {
+        checkArgument(batchMessagesMaxMessagesPerBatch >= 0, "batchMessagesMaxMessagesPerBatch needs to be >= 0");
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Why this can’t be negative if batching is disabled?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 1 above)
Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Firstly, thanks all for the review.

There is the similar check _deprecated_ `ProducerConfiguration.setBatchingMaxPublishDelay` and could not see any use-case for negative access on code-base (**e.g:** Unit Test case)
In general, i was thinking that this kind of validation can be useful due to _public API_. However, if there is valid use-case, it can be skipped and UT can be useful to highlight negative case.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 2 above)
Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: There are the similar checks on following flows:
- _deprecated_ `ProducerConfiguration.setBatchingMaxMessages` 
- `ProducerConfiguration.cc` => `setBatchingMaxMessages`

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 3 above)
Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: There is the similar check on following flows:
- _deprecated_ `ProducerConfiguration.setBatchingMaxMessages` 
- `ProducerConfiguration.cc` => `setBatchingMaxMessages`

Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: if ProducerConfiguration already verifies it, why do we need to verify it in the producer builder? The implementation of builders are wrappers for configuration objects.

Also ProducerConfiguration.setBatchingMaxMessages is verifying `> 0`. but you are adding one for `>= 0`. This is a problem, no?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 1 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: if ProducerConfiguration already verifies it, why do we need to the verification in builder? Also `ProducerConfiguration.setBatchingMaxPublishDelay` uses `>= 1` as the verification rule, but you are using `>=0`. what is the thought behind this?

Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Thanks for review and sorry for the late reply.

`ProducerConfigurationData` is being wrapped by `ProducerBuilderImpl` and `ProducerConfiguration`. However, i have mentioned `ProducerConfiguration`, not `ProducerConfigurationData`.

So, my point is that these validations exist on **deprecated** `ProducerConfiguration` API and not exist **new** `ProducerBuilderImpl` API. Does it make sense to align both of them as latest submitted patch?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 3 above)
Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Same comment as above.

Also, there is following mismatch on **legacy C++ and Java Client APIs**. Alignment can also be useful here if makes sense.

`
ProducerConfiguration.cc ## setBatchingMaxMessages => "batchingMaxMessages needs to be greater than 1"
ProducerConfiguration.setBatchingMaxMessages => checkArgument(batchMessagesMaxMessagesPerBatch > 0);
`

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 1 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: If ProducerConfigurationData already does the verification, I don't think we need to redo the verification at ProducerBuilderImpl. We should do the verification at the common class which already exists in ProducerConfigurationData. so it might be making more sense to remove the verification from deprecated ProducerConfiguration if ProducerConfigurationData already did it.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 3 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Same comments as above.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 1 above)
Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: I agree to keep these kind of validation in single tier and support required APIs by avoiding code duplication. 

So, currently, the following `setters` look like requiring alignment on `ProducerConfigurationData`.
```
setProducerName
setSendTimeoutMs
setMaxPendingMessages
setMaxPendingMessagesAcrossPartitions
setBatchingMaxPublishDelayMicros
setBatchingMaxMessages
```
Will be submitting new patch if we do not have any concern.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 3 above)
Comment (erenavsarogullari) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: Sure, will be addressing this as well.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java (hunk 1 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java]: sgtm

File: README.md [hunk 4]
```diff
@@ -67,7 +67,7 @@ $ mvn test
 Run Individual Unit Test:
 
 ```bash
-cd related-module (e.g: pulsar-client)
+$ cd related-module (e.g: pulsar-client)
```
Comment (sijie) [inline README.md]: nit: it would be great to put this change in different pull request. since it is an unrelated change.

File: pulsar-client-1x-base/pulsar-client-1x/src/main/java/org/apache/pulsar/client/api/ProducerConfiguration.java [hunk 5]
```diff
@@ -101,7 +101,7 @@ public long getSendTimeoutMs() {
      *            the time unit of the {@code sendTimeout}
      */
     public ProducerConfiguration setSendTimeout(int sendTimeout, TimeUnit unit) {
-        checkArgument(sendTimeout >= 0);
+        checkArgument(sendTimeout >= 0, "sendTimeout needs to be >= 0");
```
Comment (sijie) [inline pulsar-client-1x-base/pulsar-client-1x/src/main/java/org/apache/pulsar/client/api/ProducerConfiguration.java]: it seems that you moved the validations to ProducerConfigurationData for other methods. doesn't it make sense to move this method as well?

Comment (erenavsarogullari) [reply pulsar-client-1x-base/pulsar-client-1x/src/main/java/org/apache/pulsar/client/api/ProducerConfiguration.java]: Thanks for review. The following functions require `TimeUnit` as well (apart from the others). I agree and addressed at last patch.
```
sendTimeout(int sendTimeout, @NonNull TimeUnit unit)
batchingMaxPublishDelay(long batchDelay, @NonNull TimeUnit timeUnit)
```

File: README.md (hunk 4 above)
Comment (erenavsarogullari) [reply README.md]: Sure, will be submitting separate PR for this.

**Comment by erenavsarogullari** [comment]
run Integration tests

**Comment by erenavsarogullari** [comment]
run Integration tests

**Comment by erenavsarogullari** [comment]
Thanks all again for the reviews.
I fixed the broken UTs and PR is ready to re-review & merge.
